### PR TITLE
[`ruff_linter`] - Use LibCST in `adjust_indentation` for mixed whitespace

### LINKED ISSUE #12707: RET505 doesn't work with mixed space and tabs
ruff 0.5.6+419 (0b4d3ce39 2024-08-05)
```
ruff check *.py --select RET505 --no-cache --fix --preview --output-format concise --isolated
```

file content(at the bottom should be attached raw, not formatted file - github removes some non-printable characters, so copying from here may not work):
```
def f():
	if true:
	 return true
	else:
	 return false
```

error
```
python_data:4:2: RET505 Unnecessary `else` after `return` statement
Found 1 error.
[*] 1 fixable with the --fix option.

error: Fix introduced a syntax error in `crash` with rule codes RET505: unindent does not match any outer indentation level at byte range 33..34
---
 return falsern true
---

```

[python_code.py.zip](https://github.com/user-attachments/files/16505463/python_code.py.zip)

### PR DESCRIPTION
**Pull request by diceroll123**
## Summary

The `dedent_to` function has a difficult time with mixed whitespace.

The code now conditionally uses the LibCST method if there is mixed whitespace.

Fixes #12707

## Test Plan

`cargo test`

### CODE REVIEW

**Comment by MichaReiser** [comment]
This seems like a pretty big regression. I think I would prefer just to disable the fix in case there's mixed-indentation than regressing performance by that much.

**Comment by diceroll123** [comment]
Yeah, that is not ideal as is.

I'd like to try one more alternative approach, which does already include detecting mixed indentation.

File: crates/ruff_linter/src/fix/edits.rs [hunk 1]
```diff
@@ -302,9 +302,23 @@ pub(crate) fn adjust_indentation(
...
+        }
+
+        let line_indentation: &str = &line[..line.len() - trimmed.len()];
+        line_indentation.contains('\t') && line_indentation.contains(' ')
+    });
+
+    if mixed_indentation
+        || indexer.multiline_ranges().intersects(range)
+        || indexer.fstring_ranges().intersects(range)
+    {
```
Comment (MichaReiser) [inline crates/ruff_linter/src/fix/edits.rs:321]: The multiline comment is now misplaced

I would probably do:

```rust
    let contents = locator.slice(range);

    // If the range includes a multi-line string, use LibCST to ensure that we don't adjust the
    // whitespace _within_ the string.
		let contains_multiline_string = indexer.multiline_ranges().intersects(range)
        || indexer.fstring_ranges().intersects(range);

		let mixed_indentation = ...

		if mixed_indentation || contains_multiline_string {
			...
		}
```

**Review by MichaReiser — APPROVED** [review]
Nice!

**Pull request merged**